# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/RetainedDisplayListBuilder.cpp
# Commit: 3baead3e079b
# Full Hash: 3baead3e079bc90e7637f7f338a185149a58aa3e
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2022-02-22 09:37:09
# Regressor Bug: 1714584
# File Overlap Count: 1
# Description:
#   Bug 1714584 - Part 2: Remove nsDisplayList::RemoveBottom() r=mstange
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D138153
# ==============================================================================

diff -r a2da895a58ce -r 3baead3e079b layout/painting/RetainedDisplayListBuilder.cpp
--- a/layout/painting/RetainedDisplayListBuilder.cpp	Tue Feb 22 00:44:25 2022 +0000
+++ b/layout/painting/RetainedDisplayListBuilder.cpp	Tue Feb 22 00:44:25 2022 +0000
@@ -203,10 +203,8 @@
       aList->mDAG.Length() ==
           (initializeOldItems ? aList->Length() : aList->mOldItems.Length()));
 
-  nsDisplayList out;
-
   size_t i = 0;
-  while (nsDisplayItem* item = aList->RemoveBottom()) {
+  for (nsDisplayItem* item : aList->TakeItems()) {
 #ifdef MOZ_DIAGNOSTIC_ASSERT_ENABLED
     item->SetMergedPreProcessed(false, true);
 #endif
@@ -343,16 +341,13 @@
       if (item->GetType() == DisplayItemType::TYPE_SUBDOCUMENT) {
         IncrementSubDocPresShellPaintCount(item);
       }
-      out.AppendToTop(item);
+      aList->AppendToTop(item);
     }
     i++;
   }
 
   MOZ_RELEASE_ASSERT(aList->mOldItems.Length() == aList->mDAG.Length());
 
-  if (aKeepLinked) {
-    aList->AppendToTop(&out);
-  }
   return true;
 }
 
@@ -856,7 +851,7 @@
   MergeState merge(this, *aOldList, aOuterItem);
 
   Maybe<MergedListIndex> previousItemIndex;
-  while (nsDisplayItem* item = aNewList->RemoveBottom()) {
+  for (nsDisplayItem* item : aNewList->TakeItems()) {
     Metrics()->mNewItems++;
     previousItemIndex = merge.ProcessItemFromNewList(item, previousItemIndex);
   }
@@ -1561,9 +1556,7 @@
 void CollectStackingContextItems(nsDisplayListBuilder* aBuilder,
                                  nsDisplayList* aList, nsIFrame* aOuterFrame,
                                  int aDepth = 0, bool aParentReused = false) {
-  nsDisplayList out;
-
-  for (nsDisplayItem* item : *aList) {
+  for (nsDisplayItem* item : aList->TakeItems()) {
     if (DL_LOG_TEST(LogLevel::Debug)) {
       DL_LOGD(
           "%*s Preprocessing item %p (%s) (frame: %p) "
@@ -1600,7 +1593,7 @@
     if (aParentReused) {
       // Keep the contents of the current container item linked.
       RDLUtils::AssertDisplayItemUnmodified(item);
-      out.AppendToTop(item);
+      aList->AppendToTop(item);
     } else if (isStackingContextItem) {
       // |item| is a stacking context item that can be reused.
       ReuseStackingContextItem(aBuilder, item);
@@ -1615,9 +1608,6 @@
       IncrementPresShellPaintCount(aBuilder, item);
     }
   }
-
-  aList->Clear();
-  aList->AppendToTop(&out);
 }
 
 }  // namespace RDL